src/App.tsx

```diff
@@ -1,61 +1,154 @@
+import { useState, FormEvent } from 'react';
+
 import './App.scss';
 
-// import usersFromServer from './api/users';
-// import todosFromServer from './api/todos';
+import { TodoList } from './components/TodoList';
+import { User } from './types/User';
+import { Todo } from './types/Todo';
+
+import usersFromServer from './api/users';
+import todosFromServer from './api/todos';
+
+function getUser(userId: number): User | null {
+  return usersFromServer.find((user) => user.id === userId) || null;
+}
+
+export const todos: Todo[] = todosFromServer.map((todo) => ({
+  ...todo,
+  user: getUser(todo.userId),
+}));
+
+const allowedChars = `0123456789
+  abcdefghigklmnopqrstuvwxyz
+  абвгдеёжзийклмнопрстуфхцчшщъыьэюяґєії`;
+
+type NewTodo = {
+  title: string;
+  userName: string;
+  isTitleValid: boolean;
+  isUserNameValid: boolean;
+};
+
+const emptyTodo: NewTodo = {
+  title: '',
+  userName: '',
+  isTitleValid: true,
+  isUserNameValid: true,
+};
+
+let nextTodoId = Math.max(...todosFromServer.map((todo) => todo.id)) + 1;
+
+export const App: React.FC = () => {
+  const [todoList, setTodoList] = useState<Todo[]>(todos);
+
+  const [newTodo, setNewTodo] = useState<NewTodo>(emptyTodo);
+
+  const handleChange = (
+    e: React.ChangeEvent<HTMLInputElement | HTMLSelectElement>,
+  ) => {
+    const { name, value } = e.target;
+
+    if (
+      name === 'title'
+      && allowedChars.includes(value.slice(-1).toLowerCase())
+    ) {
+      setNewTodo((todo) => ({
+        ...todo,
+        [name]: value,
+        isTitleValid: true,
+      }));
+    }
+
+    if (name === 'userName') {
+      setNewTodo((todo) => ({
+        ...todo,
+        [name]: value,
+        isUserNameValid: true,
+      }));
+    }
+  };
+
+  const handleAddTodo = (e: FormEvent<HTMLFormElement>) => {
+    e.preventDefault();
+
+    if (!newTodo.title.trim() || !newTodo.userName) {
+      return setNewTodo((todo) => ({
+        ...todo,
+        isTitleValid: !!newTodo.title.trim(),
+        isUserNameValid: !!newTodo.userName,
+      }));
+    }
+
+    const todoUser: User | undefined = usersFromServer.find(
+      (user) => user.name === newTodo.userName,
+    );
+
+    const todoToAdd: Todo = {
+      id: nextTodoId,
+      title: newTodo.title.trim(),
+      userId: todoUser?.id,
+      completed: false,
+      user: todoUser,
+    };
+
+    setTodoList((prevTodos) => [...prevTodos, todoToAdd]);
+
+    nextTodoId += 1;
```

Comment:
Replace 'return setNewTodo(emptyTodo);' with 'setNewTodo(emptyTodo);' because the return statement is unnecessary in this case.